### PR TITLE
Use PSR-18 (and PSR-17) interfaces instead of PHP-HTTP

### PR DESCRIPTION
**Pull request by JaZo**
## Description

I updated the client so it requires a PSR-18 HTTP client instead of a PHP-HTTP client. As this is a breaking change, I took the liberty to also change the way our client is bound in the service provider. The client now discovers HTTP clients itself. A custom HTTP client must now be registered by customizing the container binding.

## Motivation and context

PSR-18 has been released for a while and most PHP-HTTP implementations now also implement PSR-18. Both changes should also make the usage without Laravel easier.

I also implemented PSR-17 HTTP factories as the currently used factory in combination with the discovery is deprecated. Currently just a few libraries implement these interfaces, so I required [nyholm/psr7](https://github.com/Nyholm/psr7) for now. Once most HTTP clients implement these interfaces (Guzzle for example is working on it), we can drop it as a dependency and leave it to the consumer to choose an implementation.

## How has this been tested?

Tested with existing/updated unit tests.

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

Go over all the following points, and put an `x` in all the boxes that apply.

- [x] I have read the **[CONTRIBUTING](CONTRIBUTING.md)** document.
- [x] My pull request addresses exactly one patch/feature.
- [x] I have created a branch for this patch/feature.
- [x] Each individual commit in the pull request is meaningful.
- [x] I have added tests to cover my changes.
- [x] If my change requires a change to the documentation, I have updated it accordingly.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -7,15 +7,20 @@
         "ext-json": "*",
         "illuminate/support": "5.5.*|5.6.*|5.7.*|5.8.*",
         "jenssegers/model": "^1.1",
-        "php-http/client-implementation": "^1.0",
-        "php-http/discovery": "^1.0"
+        "nyholm/psr7": "^1.2",
```
Comment (bbrala) [inline composer.json]: Why is this included? I do not see any usages of this namespace in the package. I do see it as a dependency for the discovery i guess, but why would you include it in the package root?

File: src/Client.php [hunk 2]
```diff
@@ -180,18 +198,13 @@ protected function getEndpoint(string $endpoint): string
         return $this->baseUri.$endpoint;
     }
 
-    protected function mergeHeaders(array $headers = []): array
+    /**
+     * @param array $headers
+     *
+     * @return array
+     */
+    protected function mergeHeaders(array $headers): array
```
Comment (bbrala) [inline src/Client.php]: Any reason to remove the default value here? Other than it being kind of useless since looking at what calls it.

File: composer.json (hunk 1 above)
Comment (JaZo) [reply composer.json]: Please see "Motivation and context".

File: src/Client.php (hunk 2 above)
Comment (JaZo) [reply src/Client.php]: Just because it's useless.

File: composer.json (hunk 1 above)
Comment (bbrala) [reply composer.json]: Fair enough, didn't see that part appearantly.
